### PR TITLE
c-ares: upgrade library to 1.20.1

### PR DESCRIPTION
**Pull request by deveshkandpal1224**
Commit Message: Upgrade c-ares library to 1.20.0.
Additional Description: This change does following:
- Updates c-ares library to 1.20 to get access to the [required](https://github.com/c-ares/c-ares/pull/549) change in c-ares library.
- Updates dirty_channel status check logic to conditionally check for status == `ARES_EREFUSED` without which `DestroyChannelOnRefused` and `CustomResolverValidAfterChannelDestruction` test fail after upgrading the c-ares library to 1.20.1.
Risk Level: low
Testing: unit testing
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: n/a
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: changelogs/current.yaml
Comment (phlax) [inline changelogs/current.yaml:27]:
```suggestion
    Upgraded c-ares library to 1.20 and added fix to c-ares DNS implementation to additionally check for ``ARES_EREFUSED`` status.
```

File: changelogs/current.yaml
Comment (phlax) [inline changelogs/current.yaml:28]:
```suggestion
    Without this fix, ``DestroyChannelOnRefused`` and ``CustomResolverValidAfterChannelDestruction`` unit test will break.
```

**Review by phlax — COMMENTED** [review]
thanks @deveshkandpal1224 - a couple of changelog nits

**Comment by deveshkandpal1224** [comment]
@phlax / @htuch  - addressed comments. Can you PTAL when you have a moment ? TIA!

File: changelogs/current.yaml
Comment (deveshkandpal1224) [reply changelogs/current.yaml]: fixed.

File: changelogs/current.yaml
Comment (deveshkandpal1224) [reply changelogs/current.yaml]: fixed.

File: bazel/repository_locations.bzl
Comment (moderation) [inline bazel/repository_locations.bzl:226]: The latest `c-ares` is `1.28.1` - https://github.com/c-ares/c-ares/releases/tag/cares-1_28_1

This now closed PR outlines some of the changes required. It would be good to catch up this dependency to the latest

Comment (deveshkandpal1224) [reply bazel/repository_locations.bzl:226]: @moderation  / @htuch  - is it alright if we upgrade to `1.28.1` with a follow up PR. Intention to upgrade to 1.20 was to help unblock [this](https://github.com/envoyproxy/envoy/pull/33551) PR for which 1.20 is the min requirement .

File: bazel/repository_locations.bzl
Comment (yanavlasov) [inline bazel/repository_locations.bzl:221]: 1.20.0 has a crashing bug. Please update to at least 1.20.1 where it was fixed.

File: source/extensions/network/dns_resolver/cares/dns_impl.cc
Comment (yanavlasov) [inline source/extensions/network/dns_resolver/cares/dns_impl.cc:197]: c-ares used to report 3 responses as ARES_ECONNREFUSED, which now get bubbled up as the original response codes. https://github.com/c-ares/c-ares/blob/ebb0a631ef10c88478e0da4eb5176807bfa0ef54/src/lib/ares_process.c#L711

I think this needs to check for all 3 status now: ARES_ESERVFAIL, ARES_ENOTIMP and ARES_EREFUSED

**Review by yanavlasov — CHANGES_REQUESTED** [review]
/wait

**Comment by deveshkandpal1224** [comment]
thanks for the review comments @yanavlasov - let me fix the condition check and upgrade the library to 1.20.1 today.

**Comment by deveshkandpal1224** [comment]
@yanavlasov - when you have a moment, can you PTAL at changes ? TIA!

**Review by yanavlasov — APPROVED** [review]

**Comment by yanavlasov** [comment]
I'm ok with this incremental upgrade. c-ares have changes that break its API behavior and jumping all the way to 1.28 may cause issues for anyone who needs to upgrade it internally (I know it does for us).

So an incremental upgrade to allow internal repos to catch up would be better in this case.

Waiting for message from deps reviewer.

/wait-any

**Comment by deveshkandpal1224** [comment]
@mattklein123 - looks like its waiting for your review ? When you have a moment, could you PTAL ? TIA!

**Comment by deveshkandpal1224** [comment]
@yanavlasov - PR seems to be stuck on @mattklein123 's approval. Wasn't sure if 2 approvals was necessary ( apologies for asking silly questions, making changes in this area for the first time :) )
 
@alyssawilk is this something you can help with ( PR approval / merge ) ? 

@phlax - sorry wasn't sure if that was more of an FYI that there will be another upgrade coming in and we are okay to proceed with 1.20.1 for now ( incremental upgrade ) so that my other [PR](https://github.com/envoyproxy/envoy/pull/33551/files) can be unblocked or were you suggesting that we should wait ?

**Review by htuch — APPROVED** [review]

**Comment by zirain** [comment]
this seems broken istio

**Comment by ramaraochavali** [comment]
What did it break in Istio?

**Comment by zirain** [comment]
see the logs https://storage.googleapis.com/istio-prow/pr-logs/pull/istio_istio/50698/integ-telemetry-discovery_istio/1784409911019442176/artifacts/telemetry-tracing-otelcollector-/TestProxyTracingOpenCensusMeshConfig/_test_context/echo-1-91597-state336912947/primary-0/client-cluster-0-v1-68767d8c77-cq9xg_istio-proxy.log

**Comment by ramaraochavali** [comment]
Seems caused by this https://github.com/c-ares/c-ares/issues/571. @zirain  do you know if open telemetry exporter is a gRPC app?

**Comment by zirain** [comment]
> Seems caused by this [c-ares/c-ares#571](https://github.com/c-ares/c-ares/issues/571). @zirain do you know if open telemetry exporter is a gRPC app?

opencensus receiver is a gRPC app.

**Pull request merged**
